sim/pkg/model/replicas_active.go

```diff
@@ -40,31 +41,52 @@ func (ras *replicasActiveStock) Count() uint64 {
 	return ras.delegate.Count()
 }
 
-func (ras *replicasActiveStock) EntitiesInStock() []*simulator.Entity {
+func (ras *replicasActiveStock) EntitiesInStock() map[simulator.Entity]bool {
 	return ras.delegate.EntitiesInStock()
 }
 
-func (ras *replicasActiveStock) Remove() simulator.Entity {
-	entity := ras.delegate.Remove()
-	if entity == nil {
+func (ras *replicasActiveStock) GetEntityByNumber(number int) simulator.Entity {
+	return *ras.replicas[number]
+}
+
+func (ras *replicasActiveStock) Remove(entity *simulator.Entity) simulator.Entity {
+	removedEntity := ras.delegate.Remove(entity)
+	if removedEntity == nil {
 		return nil
 	}
 
-	replica := entity.(Replica)
+	replica := removedEntity.(ReplicaEntity)
 	replica.Deactivate()
 
-	return entity
+	//support replicas array updated
+	ras.deleteReplica(&replica)
+	return removedEntity
 }
 
 func (ras *replicasActiveStock) Add(entity simulator.Entity) error {
-	replica := entity.(Replica)
+	replica := entity.(ReplicaEntity)
 	replica.Activate()
+	ras.replicas = append(ras.replicas, &replica)
 	return ras.delegate.Add(entity)
 }
 
+func (ras *replicasActiveStock) deleteReplica(replica *ReplicaEntity) {
```

Comment:
The `EntitiesInStock` method returns an entity set (`map[simulator.Entity]bool`).  But internally you store it as a slice.  Which makes it a little awkward because here you have to iterate through the whole list instead of just looking up the entity.

I would almost rather just continue returning a slice of `*simulatorEntity`.  Is the performance improvement of using a map instead of a slice enough to warrant the additional complexity (conversion, confusion, etc..)?  E.g. why does the map return `simulator.Entity` literal instead of a pointer, while other methods take a `*simulator.Entity`?

Reply:
Chatted with Julia.  We agreed that there should be 2 underlying implementations of the stock, one optimized for homogeneous entities and one optimized for heterogeneous entities.  The difference will be the runtime of the `Add` `Delete` and `GetAll` methods.  Additionally, the homogeneous stock base will promise stability for the `GetAll` method (which will continue to return a slice) so the get by number method won't be needed for round-robin routing.